Sergeev.georgii

sergeevgk opened the pull request.

sergeevgk committed:
Lab1 linux daemon

Add lab1 'linux daemon'

sergeevgk committed:
Lab1 linux daemon

Add requested changes

sergeevgk committed:
Lab2 host-client app

Add lab2 ver.1

KotsMikhail reviewed (changes requested):
Ходы волка и козлят должны быть в обратном порядке. Сначала козлята прячутся - потом волк их ловит.
Неадекватное поведение хоста, в случае, если в качестве хода волка ввести строку
Как получилось так, что ваш код процентов на 50 совпадает с кодом Жукова?

sergeevgk committed:
Fix requested changes

Change turns order
Add std::cin check for incorrect input

sergeevgk commented:
Исправил замечания.
Схожесть, наверное, может проявляться в алгоритме работы хоста с клиентами. Было сложно сразу понять, как они должны взаимодействовать по заданию, и тот вариант, который есть сейчас, представляется мне единственно возможным.  Он возник в ходе обсуждения с другими студентами, без списывания.

KotsMikhail commented on sergeev.georgii/lab2/connections/conn_mq.cpp:
```diff
@@ -0,0 +1,36 @@
+#include <map>
+
+std::map<int, mqd_t> queues;
+std::map<int, const char*> names;
```
Почему нельзя обойтись без глобальных переменных в реализации видов взаимодействий?

sergeevgk replied:
Эти переменные оставались с прошлых версий, где без них не работала программа. 
Исправил, переменные убраны

KotsMikhail commented on sergeev.georgii/lab2/host/wolf.h:
```diff
@@ -0,0 +1,35 @@
+    static std::map<conn*, sem_t*> connections;
+    static std::map<conn*, Status> states;
+    static std::map<conn*, int> goats_id;
+    static std::map<conn*, message> goats_messages;
```
Если у вас в классе куча map ов с разной информацией, то, возможно. эта информация должна быть инкапсулирована где-то ещё. И почему все эти переменные должны быть статическими?

sergeevgk replied:
Исправил:
Часть информации перенесена в класс клиента. То, что осталось (std::map clients и std::map clients_messages) обязано быть static, потому что используется в методах, передаваемых в pthread_create по указателю, а он не может принимать указатель на не static функции (насколько я понял).

KotsMikhail replied:
Может, почитайте про указатели на методы

KotsMikhail commented on sergeev.georgii/lab2/interfaces/Message.h:
```diff
@@ -0,0 +1,29 @@
+
+#include <stdint.h>
+
+enum Status {
```
Лучше использовать enum class

sergeevgk replied:
Исправлено

KotsMikhail commented on sergeev.georgii/lab2/interfaces/connection_info.h:
```diff
@@ -0,0 +1,15 @@
+#include <semaphore.h>
+#ifndef LAB2_CONNECTION_INFO_H
+#define LAB2_CONNECTION_INFO_H
+struct connection_info
```
Единственная причина для существования этого класса - передать 3 параметра в конструктор клиента. Это можно сделать просто передав значения

sergeevgk replied:
Удалил класс, передаю параметры в конструкторе goat

KotsMikhail commented on sergeev.georgii/lab2/host/wolf.cpp:
```diff
@@ -0,0 +1,208 @@
+    while (true) {
+        // work-read
+        std::list <pthread_t> tids;
+        for (auto iter : connections) {
```
Возможно, распараллеливание обработки клиентов стоит вынести в отдельную функцию, иначе слишком много дублирования кода

sergeevgk replied:
Выделил работу с потоками в отдельную функцию

KotsMikhail reviewed (changes requested):
В дополнение к тому, что я написал, загляните в пул реквест Жукова. У вас есть почти все те же ошибки, что и у него, их следует исправить

sergeevgk committed:
Lab2 fix #2

Delete connection_info class -> pass client information via ctor
Add errors check
Add new wolf:: functions for code clearness
Change enum Status & Owner to enum class
Change connections implementation & delete global variables

sergeevgk commented:
Замечания к моему коду и замечания, которые я увидел к предпоследнему коммиту в ветке alexander.zhukov, исправлены

KotsMikhail reviewed (changes requested):
Дописал Жукову ещё несколько замечаний, большинство из них у вас тоже есть

KotsMikhail commented on sergeev.georgii/lab2/host/wolf.h:
```diff
@@ -0,0 +1,35 @@
+    static int num_goats;
+    static int num;
+    static const int TIMEOUT = 5;
+    static constexpr double WOLF_WORK_WINDOW = 1;
```
Чем constexpr отличается от const, и почему тут должно быть именно первое, а предыдущая константа просто const?

sergeevgk replied:
Я долго искал ответ на вопрос, почему компилятор не даёт инициализировать static сonst double внутри класса. Насколько я понял, такая инициализация работает только для т.н. "integral types" (int, char, short, long, bool). 
constexpr создает объект на этапе компиляции и использует его как константу времени компиляции. Если переменная объявлена с constexpr, то она должна быть не только создана, но и проинициализирована на этапе компиляции, в отличие от const. 
Предыдущую константу тоже можно сделать constexpr, мне почему-то не пришло в голову написать одинаково тут.

sergeevgk committed:
Lab2 fix 3

Fix memory leaks
Replace two maps(clients, messages) with list and map

sergeevgk commented:
Замечания исправлены

SWAGLORD1 commented:
Ещё одну утечку памяти вы так и не нашли, можете на неё в пул реквесте Жукова посмотреть. 7 баллов за задачу
